<a href="https://colab.research.google.com/github/JuliaCAlmeida/MeuSegundoProjeto/blob/main/Atividade_30_09.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [3]:
import numpy as np

In [7]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [4]:
dt = pd.read_csv('/content/Extended_Employee_Performance_and_Productivity_Data.csv')
dt

Employee_ID        Department  Gender  Age   Job_Title  \
0                1                IT    Male   55  Specialist   
1                2           Finance    Male   29   Developer   
2                3           Finance    Male   55  Specialist   
3                4  Customer Support  Female   48     Analyst   
4                5       Engineering  Female   36     Analyst   
...            ...               ...     ...  ...         ...   
99995        99996           Finance    Male   27  Technician   
99996        99997                IT  Female   36  Consultant   
99997        99998        Operations    Male   53     Analyst   
99998        99999                HR  Female   22  Consultant   
99999       100000           Finance  Female   43     Analyst   

                        Hire_Date  Years_At_Company Education_Level  \
0      2022-01-19 08:03:05.556036                 2     High School   
1      2024-04-18 08:03:05.556036                 0     High School   
2      2015-10-26 08:03:05.556036                 8     High School   
3      2016-10-22 08:03:05.556036                 7        Bachelor   
4      2021-07-23 08:03:05.556036                 3        Bachelor   
...                           ...               ...             ...   
99995  2022-12-07 08:03:05.556036                 1        Bachelor   
99996  2018-07-24 08:03:05.556036                 6          Master   
99997  2015-11-24 08:03:05.556036                 8     High School   
99998  2015-08-03 08:03:05.556036                 9     High School   
99999  2024-03-04 08:03:05.556036                 0             PhD   

       Performance_Score  Monthly_Salary  Work_Hours_Per_Week  \
0                      5          6750.0                   33   
1                      5          7500.0                   34   
2                      3          5850.0                   37   
3                      2          4800.0                   52   
4                      2          4800.0                   38   
...                  ...             ...                  ...   
99995                  4          4900.0                   55   
99996                  5          8250.0                   39   
99997                  2          4800.0                   31   
99998                  5          8250.0                   35   
99999                  1          4400.0                   51   

       Projects_Handled  Overtime_Hours  Sick_Days  Remote_Work_Frequency  \
0                    32              22          2                      0   
1                    34              13         14                    100   
2                    27               6          3                     50   
3                    10              28         12                    100   
4                    11              29         13                    100   
...                 ...             ...        ...                    ...   
99995                46               5          3                     75   
99996                35               7          0                      0   
99997                13               6          5                      0   
99998                43              10          1                     75   
99999                43              27         11                     75   

       Team_Size  Training_Hours  Promotions  Employee_Satisfaction_Score  \
0             14              66           0                         2.63   
1             12              61           2                         1.72   
2             10               1           0                         3.17   
3             10               0           1                         1.86   
4             15               9           1                         1.25   
...          ...             ...         ...                          ...   
99995         16              48           2                         1.28   
99996         10              77           1       

In [6]:
dt['Resigned'] = dt['Resigned'].astype(int)
dt

Employee_ID        Department  Gender  Age   Job_Title  \
0                1                IT    Male   55  Specialist   
1                2           Finance    Male   29   Developer   
2                3           Finance    Male   55  Specialist   
3                4  Customer Support  Female   48     Analyst   
4                5       Engineering  Female   36     Analyst   
...            ...               ...     ...  ...         ...   
99995        99996           Finance    Male   27  Technician   
99996        99997                IT  Female   36  Consultant   
99997        99998        Operations    Male   53     Analyst   
99998        99999                HR  Female   22  Consultant   
99999       100000           Finance  Female   43     Analyst   

                        Hire_Date  Years_At_Company Education_Level  \
0      2022-01-19 08:03:05.556036                 2     High School   
1      2024-04-18 08:03:05.556036                 0     High School   
2      2015-10-26 08:03:05.556036                 8     High School   
3      2016-10-22 08:03:05.556036                 7        Bachelor   
4      2021-07-23 08:03:05.556036                 3        Bachelor   
...                           ...               ...             ...   
99995  2022-12-07 08:03:05.556036                 1        Bachelor   
99996  2018-07-24 08:03:05.556036                 6          Master   
99997  2015-11-24 08:03:05.556036                 8     High School   
99998  2015-08-03 08:03:05.556036                 9     High School   
99999  2024-03-04 08:03:05.556036                 0             PhD   

       Performance_Score  Monthly_Salary  Work_Hours_Per_Week  \
0                      5          6750.0                   33   
1                      5          7500.0                   34   
2                      3          5850.0                   37   
3                      2          4800.0                   52   
4                      2          4800.0                   38   
...                  ...             ...                  ...   
99995                  4          4900.0                   55   
99996                  5          8250.0                   39   
99997                  2          4800.0                   31   
99998                  5          8250.0                   35   
99999                  1          4400.0                   51   

       Projects_Handled  Overtime_Hours  Sick_Days  Remote_Work_Frequency  \
0                    32              22          2                      0   
1                    34              13         14                    100   
2                    27               6          3                     50   
3                    10              28         12                    100   
4                    11              29         13                    100   
...                 ...             ...        ...                    ...   
99995                46               5          3                     75   
99996                35               7          0                      0   
99997                13               6          5                      0   
99998                43              10          1                     75   
99999                43              27         11                     75   

       Team_Size  Training_Hours  Promotions  Employee_Satisfaction_Score  \
0             14              66           0                         2.63   
1             12              61           2                         1.72   
2             10               1           0                         3.17   
3             10               0           1                         1.86   
4             15               9           1                         1.25   
...          ...             ...         ...                          ...   
99995         16              48           2                         1.28   
99996         10              77           1       

Regressão Logistica


In [8]:
ft = ['Performance_Score', 'Monthly_Salary', 'Promotions', 'Age', 'Years_At_Company']

In [9]:
X = dt[ft]
y = dt['Resigned']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [10]:
model = LogisticRegression()
model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [11]:
y_pred = model.predict(X_test)

In [12]:
print("Acurácia:", accuracy_score(y_test, y_pred))
print("\nMatriz de Confusão:\n", confusion_matrix(y_test, y_pred))
print("\nRelatório de Classificação:\n", classification_report(y_test, y_pred))

Acurácia: 0.90145

Matriz de Confusão:
 [[18029     0]
 [ 1971     0]]

Relatório de Classificação:
               precision    recall  f1-score   support

           0       0.90      1.00      0.95     18029
           1       0.00      0.00      0.00      1971

    accuracy                           0.90     20000
   macro avg       0.45      0.50      0.47     20000
weighted avg       0.81      0.90      0.85     20000



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Balanceamento da base de dados



In [17]:
resigned_true = dt[dt['Resigned'] == True]
resigned_false = dt[dt['Resigned'] == False]

In [20]:
resigned_true_sample = resigned_true.sample(n=10000, random_state=42)
resigned_false_sample = resigned_false.sample(n=10000, random_state=42)

In [24]:
balanced_df = pd.concat([resigned_true_sample, resigned_false_sample])
balanced_df

Employee_ID        Department  Gender  Age   Job_Title  \
74069        74070        Operations  Female   27   Developer   
89980        89981                IT    Male   60    Engineer   
5308          5309                HR    Male   38  Technician   
92863        92864             Sales  Female   22   Developer   
22012        22013             Sales    Male   51  Specialist   
...            ...               ...     ...  ...         ...   
59373        59374         Marketing  Female   43   Developer   
39603        39604                HR  Female   38     Manager   
32728        32729                IT  Female   48    Engineer   
78967        78968  Customer Support    Male   34  Specialist   
20951        20952        Operations  Female   53  Consultant   

                        Hire_Date  Years_At_Company Education_Level  \
74069  2020-05-16 08:03:05.556036                 4     High School   
89980  2023-09-28 08:03:05.556036                 0        Bachelor   
5308   2016-03-08 08:03:05.556036                 8        Bachelor   
92863  2022-10-13 08:03:05.556036                 1        Bachelor   
22012  2021-07-18 08:03:05.556036                 3        Bachelor   
...                           ...               ...             ...   
59373  2017-08-21 08:03:05.556036                 7     High School   
39603  2021-08-17 08:03:05.556036                 3     High School   
32728  2017-02-05 08:03:05.556036                 7        Bachelor   
78967  2020-12-21 08:03:05.556036                 3        Bachelor   
20951  2016-11-13 08:03:05.556036                 7        Bachelor   

       Performance_Score  Monthly_Salary  Work_Hours_Per_Week  \
74069                  1          5500.0                   40   
89980                  4          8400.0                   41   
5308                   1          3850.0                   47   
92863                  4          7000.0                   34   
22012                  3          5850.0                   49   
...                  ...             ...                  ...   
59373                  3          6500.0                   55   
39603                  5          9000.0                   37   
32728                  2          7200.0                   35   
78967                  1          4950.0                   41   
20951                  3          7150.0                   56   

       Projects_Handled  Overtime_Hours  Sick_Days  Remote_Work_Frequency  \
74069                22               4         10                     50   
89980                 5              23         10                     25   
5308                 26               7         13                     25   
92863                23              28         10                     75   
22012                 8              14          0                    100   
...                 ...             ...        ...                    ...   
59373                22              11          6                      0   
39603                13              23         11                     50   
32728                42               4         14                    100   
78967                32               2          8                     75   
20951                 7              14          5                    100   

       Team_Size  Training_Hours  Promotions  Employee_Satisfaction_Score  \
74069         11              10           2                         4.93   
89980          5              78           0                         2.65   
5308          14              98           2                         2.42   
92863          2              31           1                         2.52   
22012         15              26           1                         3.03   
...          ...             ...         ...                          ...   
59373          7              72           0                         2.00   
39603          2              85           2       

In [25]:
balanced_df = balanced_df.sample(frac=1, random_state=42).reset_index(drop=True)


In [27]:
df = balanced_df
df

Employee_ID        Department  Gender  Age   Job_Title  \
0            82377  Customer Support    Male   58  Specialist   
1            44091        Operations   Other   34    Engineer   
2            38935             Legal    Male   50  Technician   
3            33383                IT  Female   32     Analyst   
4            99922  Customer Support    Male   24  Technician   
...            ...               ...     ...  ...         ...   
19995        62334             Sales  Female   33  Consultant   
19996        96159  Customer Support  Female   41   Developer   
19997        99760             Sales   Other   39  Consultant   
19998        25047         Marketing    Male   34  Specialist   
19999         1447       Engineering  Female   56     Analyst   

                        Hire_Date  Years_At_Company Education_Level  \
0      2016-07-13 08:03:05.556036                 8             PhD   
1      2020-03-13 08:03:05.556036                 4        Bachelor   
2      2020-02-09 08:03:05.556036                 4             PhD   
3      2014-10-01 08:03:05.556036                 9        Bachelor   
4      2021-01-10 08:03:05.556036                 3          Master   
...                           ...               ...             ...   
19995  2018-06-24 08:03:05.556036                 6        Bachelor   
19996  2018-05-24 08:03:05.556036                 6        Bachelor   
19997  2015-07-28 08:03:05.556036                 9          Master   
19998  2019-03-22 08:03:05.556036                 5     High School   
19999  2018-10-11 08:03:05.556036                 5          Master   

       Performance_Score  Monthly_Salary  Work_Hours_Per_Week  \
0                      1          4950.0                   45   
1                      2          7200.0                   58   
2                      2          4200.0                   56   
3                      3          5200.0                   49   
4                      2          4200.0                   34   
...                  ...             ...                  ...   
19995                  5          8250.0                   40   
19996                  1          5500.0                   39   
19997                  5          8250.0                   32   
19998                  1          4950.0                   38   
19999                  3          5200.0                   47   

       Projects_Handled  Overtime_Hours  Sick_Days  Remote_Work_Frequency  \
0                    26              23          1                     25   
1                    36              23         13                     50   
2                    30              14         12                      0   
3                     8              22          9                     75   
4                    43              16          1                     50   
...                 ...             ...        ...                    ...   
19995                30              21         10                     25   
19996                14              24          3                     25   
19997                11              27          5                     75   
19998                 9              22         13                     50   
19999                16              11          7                     75   

       Team_Size  Training_Hours  Promotions  Employee_Satisfaction_Score  \
0              9               8           2                         2.10   
1             12              70           0                         4.66   
2              7              56           2                         2.76   
3              5              40           2                         4.06   
4             19              52           2                         3.65   
...          ...             ...         ...                          ...   
19995         12              51           0                         1.21   
19996          2              33           1       

In [28]:
features = ['Performance_Score', 'Monthly_Salary', 'Promotions', 'Age', 'Years_At_Company']

In [30]:
X = df[features]
y = df['Resigned']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [31]:
model = LogisticRegression()
model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [32]:
y_pred = model.predict(X_test)

In [33]:
accuracy = accuracy_score(y_test, y_pred)
print("Acurácia:", accuracy)

Acurácia: 0.4885


In [34]:
cm = confusion_matrix(y_test, y_pred)
print("\nMatriz de Confusão:\n", cm)


Matriz de Confusão:
 [[1326  619]
 [1427  628]]


In [35]:
resigned_true = df[df['Resigned'] == True]
resigned_false = df[df['Resigned'] == False]

In [36]:
resigned_true_sample = resigned_true.sample(n=100, random_state=42)
resigned_false_sample = resigned_false.sample(n=9900, random_state=42)


In [39]:
unbalanced_df = pd.concat([resigned_true_sample, resigned_false_sample])
unbalanced_df

Employee_ID        Department  Gender  Age   Job_Title  \
12509        43063           Finance  Female   51  Technician   
9307         85913         Marketing    Male   36  Technician   
3439         81078       Engineering  Female   36  Consultant   
9426         16599         Marketing  Female   58     Analyst   
8962         79915                IT  Female   57     Manager   
...            ...               ...     ...  ...         ...   
4956         91054         Marketing  Female   50  Specialist   
14523        58448                IT    Male   33     Analyst   
6234         82513                HR    Male   25     Analyst   
16841        11439  Customer Support  Female   44     Manager   
12551        57732             Legal    Male   39     Manager   

                        Hire_Date  Years_At_Company Education_Level  \
12509  2023-02-04 08:03:05.556036                 1          Master   
9307   2020-09-02 08:03:05.556036                 4             PhD   
3439   2024-01-23 08:03:05.556036                 0        Bachelor   
9426   2017-07-23 08:03:05.556036                 7        Bachelor   
8962   2023-05-08 08:03:05.556036                 1     High School   
...                           ...               ...             ...   
4956   2021-09-02 08:03:05.556036                 3          Master   
14523  2016-02-03 08:03:05.556036                 8        Bachelor   
6234   2019-04-06 08:03:05.556036                 5        Bachelor   
16841  2020-01-21 08:03:05.556036                 4     High School   
12551  2016-10-07 08:03:05.556036                 7          Master   

       Performance_Score  Monthly_Salary  Work_Hours_Per_Week  \
12509                  3          4550.0                   49   
9307                   3          4550.0                   34   
3439                   5          8250.0                   32   
9426                   4          5600.0                   37   
8962                   1          6600.0                   33   
...                  ...             ...                  ...   
4956                   4          6300.0                   53   
14523                  5          6000.0                   41   
6234                   1          4400.0                   51   
16841                  1          6600.0                   54   
12551                  3          7800.0                   47   

       Projects_Handled  Overtime_Hours  Sick_Days  Remote_Work_Frequency  \
12509                19              22         11                     25   
9307                 15              27          9                     75   
3439                  4              10         14                     50   
9426                 39               9          5                     75   
8962                 27              28          0                      0   
...                 ...             ...        ...                    ...   
4956                 12              13          2                     50   
14523                29              23          4                     25   
6234                 24               6         10                     50   
16841                43               9          4                     75   
12551                46              15          6                      0   

       Team_Size  Training_Hours  Promotions  Employee_Satisfaction_Score  \
12509          7              60           2                         4.54   
9307          12              59           1                         3.17   
3439           5               1           0                         3.62   
9426          16              92           0                         2.13   
8962           5              34           1                         1.51   
...          ...             ...         ...                          ...   
4956          16              28           1                         2.22   
14523          6              87           2       

In [40]:
unbalanced_df = unbalanced_df.sample(frac=1, random_state=42).reset_index(drop=True)
unbalanced_df


Employee_ID   Department  Gender  Age   Job_Title  \
0           46190      Finance  Female   40   Developer   
1           98575      Finance  Female   40     Manager   
2            1280   Operations    Male   43    Engineer   
3           76718           IT  Female   26   Developer   
4           79208      Finance  Female   22     Analyst   
...           ...          ...     ...  ...         ...   
9995         6213   Operations  Female   37     Analyst   
9996        32702      Finance    Male   60  Specialist   
9997        42528   Operations    Male   37  Consultant   
9998        57049    Marketing    Male   41    Engineer   
9999        63461  Engineering  Female   57    Engineer   

                       Hire_Date  Years_At_Company Education_Level  \
0     2024-01-28 08:03:05.556036                 0     High School   
1     2020-12-17 08:03:05.556036                 3        Bachelor   
2     2016-12-15 08:03:05.556036                 7          Master   
3     2017-09-15 08:03:05.556036                 6             PhD   
4     2018-03-21 08:03:05.556036                 6          Master   
...                          ...               ...             ...   
9995  2024-07-28 08:03:05.556036                 0     High School   
9996  2018-12-22 08:03:05.556036                 5     High School   
9997  2017-05-19 08:03:05.556036                 7        Bachelor   
9998  2018-02-28 08:03:05.556036                 6     High School   
9999  2014-09-20 08:03:05.556036                 9     High School   

      Performance_Score  Monthly_Salary  Work_Hours_Per_Week  \
0                     3          6500.0                   31   
1                     1          6600.0                   32   
2                     4          8400.0                   47   
3                     1          5500.0                   39   
4                     1          4400.0                   47   
...                 ...             ...                  ...   
9995                  5          6000.0                   52   
9996                  2          5400.0                   47   
9997                  1          6050.0                   30   
9998                  4          8400.0                   47   
9999                  4          8400.0                   38   

      Projects_Handled  Overtime_Hours  Sick_Days  Remote_Work_Frequency  \
0                   49               7          9                     75   
1                   31              24         10                     25   
2                   16              18          8                     25   
3                   38               1          0                      0   
4                    6              14         11                     75   
...                ...             ...        ...                    ...   
9995                 0               0          2                    100   
9996                27              11          7                     50   
9997                26              18         14                    100   
9998                 3              19          2                     25   
9999                30              19          1                    100   

      Team_Size  Training_Hours  Promotions  Employee_Satisfaction_Score  \
0             7              31           0                         2.73   
1            13              37           2                         1.76   
2             4               4           1                         4.46   
3            12              99           0                         1.72   
4             5              14           2                         3.13   
...         ...             ...         ...                          ...   
9995          3              78           0                         1.33   
9996         15              89           0                         1.29   
9997          3              51           1                         3.25   
9998         1

In [41]:
udf = unbalanced_df

Recalcule a Regressão Logística

In [42]:
X = df[features]
y = df['Resigned']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [43]:
X = df[features]
y = df['Resigned']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [44]:
y_pred = model.predict(X_test)

In [45]:
accuracy = accuracy_score(y_test, y_pred)
print("Acurácia:", accuracy)


Acurácia: 0.4885


In [46]:
cm = confusion_matrix(y_test, y_pred)
print("\nMatriz de Confusão:\n", cm)


Matriz de Confusão:
 [[1326  619]
 [1427  628]]


Compare as matrizes de confusão e a acurácia. O que pode ser observado nessa comparação? O que a matriz de confusão nos diz em cada um dos 3 modelos treinados?

Modelo 01:

Acurácia: 0.90145

Matriz de Confusão:
 [[18029     0]
 [ 1971     0]]


 Modelo 02:

Acurácia: 0.4885

Matriz de Confusão:
 [[1326  619]
 [1427  628]]


 Modelo 03:

 Acurácia: 0.4885

 Matriz de Confusão:
 [[1326  619]
 [1427  628]]